### ISSUE TITLE
cryptol -c does not return exit code

### ISSUE THREAD

**Issue by benjaminselfridge** [GaloisInc/cryptol#974]
`cryptol -c file.icry` return an exit code of `0` even if there is an error in executing the script. It should return a nonzero code to indicate failure.

**Comment by brianhuffman** [comment]
So, can we put together a complete list of the commands that can "fail"? Here's what I'm thinking so far:

* `:l` and `:m`, if the file doesn't successfully load, parse, and typecheck
* `:prove`, `:check`, or `:exhaust`, if they find counterexamples
* `:type`, if the expression has a type error
* expression evaluation, if it produces a run-time error

Are there others? Also, what about `:sat`? Does unsat mean failure?

**Comment by yav** [comment]
You can use `-e` if you want `Cryptol` to stop on error, and return an error code.

We have the distinction because in some cases finding a type error *is* the successful behavior, which happens when running test for example.

**Comment by brianhuffman** [comment]
In that case, does `cryptol -e` already do everything that @benjaminselfridge wants? I never knew about `-e`! And if I'm implementing a cryptol REPL command, how do I indicate that it should throw an error that `-e` cares about?

**Comment by benjaminselfridge** [comment]
I also didn't see the `-e` command. As far as I'm concerned, this issue can be closed -- this solves my problem!

**Comment by weaversa** [comment]
I don’t think -e works though.

**Comment by weaversa** [comment]
We tried using it in Travis CI scripts and -e always seemed to return 0

**Comment by weaversa** [comment]
```
$ cryptol -e nothere.cry -c ":check" && echo $?
[error] can't find file: nothere.cry
Loading module Cryptol
There are no properties in scope.
0
```

**Comment by weaversa** [comment]
```
$ cat p.cry && cryptol -e p.cry -c ":check" && echo $?
property a = True

property b = False

property c = True
Loading module Cryptol
Loading module Main
property a Using exhaustive testing.
Testing... Passed 1 tests.
Q.E.D.
property b Using exhaustive testing.
Testing... b = False
property c Using exhaustive testing.
Testing... Passed 1 tests.
Q.E.D.
0
```

**Comment by weaversa** [comment]
I don't know how to get `cryptol -e` to stop on error or return something other than `0`.

**Comment by weaversa** [comment]
```
$ cat p.cry && cryptol -e p.cry -c ":check" && echo $?
property a = True

property b = False

blalkdf

property c = True
Parse error at p.cry:7:9,
  unexpected: layout block separator
Loading module Cryptol
There are no properties in scope.
0
```

**Comment by brianhuffman** [comment]
Perhaps we should rename this ticket to "cryptol -e does not return exit code".

**Comment by yav** [comment]
@weaversa I think the issue is that you are using `&&`.  This works for me:
```
> cryptol -e -c '72 : X'; echo $?
```
The result is:
```
Loading module Cryptol

[error] at <interactive>:1:6--1:7 Type not in scope: X
1
```

**Comment by yav** [comment]
@brianhuffman I just looked at the code and basically `-e` cares about unhandled errors in the `REPL` monad (and some other things, but that'd be the way to signal a problem on the REPL I think).

**Comment by benjaminselfridge** [comment]
`-e` worked in the small example I tried, fwiw

**Comment by weaversa** [comment]
@yav You're right in that I should have used `;` rather than `&&`, but `cryptol -e` still does not report an error for any of the tests I tried.

```
$ cryptol -e doesnotexist.cry -c ":check" ; echo $?
[error] can't find file: doesnotexist.cry
Loading module Cryptol
There are no properties in scope.
0
```

```
$ cryptol -e p.cry -c ":check" ; echo $?
Loading module Cryptol
Loading module Main
property a Using exhaustive testing.
Testing... Passed 1 tests.
Q.E.D.
property b Using exhaustive testing.
Testing... b = False
property c Using exhaustive testing.
Testing... Passed 1 tests.
Q.E.D.
0
```

**Comment by yav** [comment]
Good to know, let's keep this ticket open until we've fixed those.  The "missing file" one in particular really should be an error.   I guess a failing test should be considered an error too.

**Comment by brianhuffman** [comment]
In order to close this ticket I think we just need to do the following:
- Decide and document which commands should be considered "failure"
- Make sure that `cryptol -e` returns the appropriate error code in each case

**Comment by WeeknightMVP** [comment]
Should `cryptol -e -c $CMD` and `$CMD > cmd.icry && cryptol -e -b cmd.icry` ever return different error codes?  (Assuming multiple `-c {command}` for a batch file with multiple commands, of course.)

```
$ echo ":l missing.cry" > missing.icry; cryptol -e -b missing.icry; echo "$?"
Loading module Cryptol

[error] can't find file: missing.cry
1

$ cryptol -e -c ":l missing.cry"; echo "$?"
Loading module Cryptol

[error] can't find file: missing.cry
0
```
